### PR TITLE
Dict space support for VecEnv

### PR DESCRIPTION
**Pull request by AdamGleave**
OpenAI added support for Dict observation spaces to DummyVecEnv and SubprocVecEnv after the fork. This PR cherry-picks these changes, while retaining the improvements made by Stable Baselines. This will be needed (amongst other changes) to support use cases such as #133.

OpenAI also added ShmemVecEnv, with common code factored out into util.py. I've retained the util.py as a separate file in case we wish to also port over ShmemVecEnv in the future.

### CODE REVIEW

**Comment by araffin** [comment]
Hello,
Thanks for this pr. Before any review, please read "CONTRIBUTING.md" carefully ;)

**Comment by AdamGleave** [comment]
Thanks @araffin. I think the only things I missed were updating the change log (done now) and creating an accompanying issue for the PR (closes #208). Let me know if there are any changes you'd like to see.

There's also a case for adding `Tuple` support, as discussed in the issue; let me know if you'd like to see this and I can add it as well.

**Comment by araffin** [comment]
>I think the only things I missed

Well, the issue was already there (see #133 ). Yes for the changelog, but I was more concern about the import and the documentation of each function.

**Comment by AdamGleave** [comment]
Thanks for the comments. Re: the imports, I assume the issue is with the relative import style. This was already present in `DummyVecEnv` so I did not want to change it, but have now switched to an absolute import style.

I've expanded on the documentation for the functions added and made them a consistent style. Happy to elaborate on this more but the functions added in the PR are all module-private so documentation doesn't seem as important as if part of public API.

**Comment by araffin** [comment]
I should have been more explicit, relative import are fine, i was refering to the `import collection` that was not at the top of the file.
Sure, go ahead if you can also add support for tuple =)!
Every function documentation matter ;)

**Comment by AdamGleave** [comment]
I've added support for Tuple observations, ready for review.

One area I'd appreciate feedback is what axis ordering we want for actions. Note there are two axes in the observations/actions for `Tuple` and `Dict` spaces in `VecEnv` environments: the index into the `tuple` or `dict`, and the index of the environment. 

For observations, the first axis indexes into the `tuple` or `dict`, and the environment id second. This follows the convention used in OpenAI master, and I think is generally a reasonable design choice: e.g. one might want to feed a `dict` of observations of differently shaped values into different placeholders in a TensorFlow graph.

However, I'm less sure what the right choice is for actions. Right now, `step_async` follows the opposite convention: first index is environment, second index is into `tuple` or `dict`. This is the same convention as in OpenAI master, and does have its advantages: e.g. random sampling is easy, `[action_space.sample() for _ in range(num_envs)]`. However, I dislike the inconsistency, and for my own application (multi-agent RL, where each tuple index corresponds to a different agent) it would be more convenient to have the `tuple` or `dict` axis come first. 

Converting between these two conventions is just a transpose operation, so the user can always change it, but it'd be good to get the default right.

**Comment by araffin** [comment]
Hello,

I did a quick test today and found a minor inconsistency:
- the implementation seems inconsistent between DummyVecEnv and SubprocVecEnv: one requires an ordered dict while the other is ok with a normal dict. Also, that is a problem because it just crashes if this requirement is not satisfied.

It also seems that the doc needs to be updated (+ adding a note saying that currently no model supports those spaces).

I'll try to have a deeper look in the coming days but I'm quite busy right now...

PS: please add your name at the bottom of the changelog too ;)

**Comment by AdamGleave** [comment]
Thanks for the review and catching the error with `dict` v.s. `OrderedDict`. Gym is a little unclear on this point: the `Dict` space has `spaces` be an `OrderedDict` and `sample` always returns an `OrderedDict`, but `contains` accepts any `dict` instance (and some Gym environments return unordered `dict`s as observations). Following Postel's law, I've modified the `VecEnv` classes to accept unordered `dict` as observations, but to always return an `OrderedDict` to the user. I've also added a test case to verify this behaviour.

What docs would you like me to update? I believe the `VecEnv` classes should now support all observation spaces supported in Gym; do you think this should be stated explicitly (currently the docs do not mention anything regarding supported observation spaces)?

**Comment by araffin** [comment]
>What docs would you like me to update?

I would say [this one](https://stable-baselines.readthedocs.io/en/master/guide/vec_envs.html). 

>do you think this should be stated explicitly (currently the docs do not mention anything regarding supported observation spaces)?

yes, that what I meant for updating the doc. Please add a note about what is supported by the vec env.
For the algorithm, we already have the [table](https://stable-baselines.readthedocs.io/en/master/guide/algos.html) in the doc and a more complete one in the readme.

File: stable_baselines/common/vec_env/util.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    Unstructured spaces are represented by {None: obs_space}.
+
+    :param obs_space: (gym.spaces.Space) an observation space
+    :return (tuple) A tuple (keys, shapes, dtypes):
+        keys: a list of dict keys.
+        shapes: a dict mapping keys to shapes.
+        dtypes: a dict mapping keys to dtypes.
+    """
+    if isinstance(obs_space, gym.spaces.Dict):
+        assert isinstance(obs_space.spaces, OrderedDict)
```
Comment (araffin) [inline stable_baselines/common/vec_env/util.py]: Please add an informative message if the assertion fails.

File: stable_baselines/common/vec_env/util.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        subspaces = obs_space.spaces
+    elif isinstance(obs_space, gym.spaces.Tuple):
+        subspaces = {i: space for i, space in enumerate(obs_space.spaces)}
+    else:
+        assert not hasattr(obs_space, 'spaces')
```
Comment (araffin) [inline stable_baselines/common/vec_env/util.py]: idem (please add a message for the user)

File: stable_baselines/common/vec_env/subproc_vec_env.py [hunk 3]
```diff
@@ -161,3 +164,27 @@ def set_attr(self, attr_name, value, indices=None):
...
+def _flatten_obs(obs, space):
+    """
+    Flatten observations, depending on the observation space.
+
+    :param obs: (dict<ndarray>, tuple<ndarray> or ndarray) an observation.
+                Either a numpy array or a dict or tuple of numpy arrays.
+    :return (OrderedDict<ndarray>, tuple<ndarray> or ndarray) a flattened numpy array
+            or an OrderedDict or tuple of flattened numpy arrays.
+    """
+    assert isinstance(obs, list) or isinstance(obs, tuple)
```
Comment (araffin) [inline stable_baselines/common/vec_env/subproc_vec_env.py]: This does not match the docstring, no?

Also, you can simplify the check:

```python
assert isinstance(obs, (list, tuple))
```

File: stable_baselines/common/vec_env/subproc_vec_env.py [hunk 4]
```diff
@@ -161,3 +164,27 @@ def set_attr(self, attr_name, value, indices=None):
...
+    assert len(obs) > 0
+
+    if isinstance(space, gym.spaces.Dict):
+        assert isinstance(obs[0], dict)
+        return collections.OrderedDict([(k, np.stack([o[k] for o in obs]))
+                                        for k in space.spaces.keys()])
```
Comment (araffin) [inline stable_baselines/common/vec_env/subproc_vec_env.py]: ok, this works because space.spaces.keys() will always be ordered. (Would not be True for python 3.5 if is was a normal dict)

File: stable_baselines/common/vec_env/util.py (hunk 1 above)
Comment (AdamGleave) [reply stable_baselines/common/vec_env/util.py]: done

File: stable_baselines/common/vec_env/subproc_vec_env.py (hunk 3 above)
Comment (AdamGleave) [reply stable_baselines/common/vec_env/subproc_vec_env.py]: good catch, updated docstring.

File: stable_baselines/common/vec_env/subproc_vec_env.py (hunk 4 above)
Comment (AdamGleave) [reply stable_baselines/common/vec_env/subproc_vec_env.py]: Yep, this is part of the interface for `gym.spaces.Dict`. I've added an assert statement to verify this and clarify it to the reader.

**Comment by AdamGleave** [comment]
Thanks for your comments. I've added documentation on supported spaces to the vector environment page, and also clarified on the RL algorithms page that the algorithms do not currently support `Dict` and `Tuple` spaces. I believe I've addressed the comments you made on the code. I also realised while writing the documentation I was missing test cases for some of the more obscure spaces (e.g. `MultiBinary`) so added those.

File: docs/guide/algos.rst [hunk 5]
```diff
@@ -37,6 +37,7 @@ TRPO         ✔️                        ✔️         ✔️           ✔
 .. [#f4] Multi Processing with `MPI`_.
 .. [#f5] TODO, in project scope.
 
+Note non-array spaces such as `Dict` or `Tuple` are not currently supported by any algorithm.
```
Comment (araffin) [inline docs/guide/algos.rst]:
```suggestion
.. note::

    Non-array spaces such as `Dict` or `Tuple` are not currently supported by any algorithm.
```

Minor: formatting for sphinx doc

File: docs/guide/vec_envs.rst [hunk 6]
```diff
@@ -5,19 +5,27 @@
...
-
+Vectorized Environments are a method for multiprocess training. Instead of training an RL agent
+on 1 environment, it allows us to train it on `n` environments using `n` processes.
+Because of this, `actions` passed to the environment are now a vector (of dimension `n`).
+It is the same for `observations`, `rewards` and end of episode signals (`dones`).
+In the case of non-array observation spaces such as `Dict` or `Tuple`, where different sub-spaces
+may have different shapes, the sub-observations are vectors (of dimension `n`).
+
+============= ============== ======= ============ ======== ========= ================
+Name          Refactored     ``Box`` ``Discrete`` ``Dict`` ``Tuple`` Multi Processing
```
Comment (araffin) [inline docs/guide/vec_envs.rst]: Here I think you can remove the "Refactored" column. And I'm not sure about the "Multi Processing" one. @hill-a  ?

**Comment by araffin** [comment]
Thank you very much for your good PR.
Just minor details in the doc, it will be merged after ;)

File: docs/guide/vec_envs.rst (hunk 6 above)
Comment (hill-a) [reply docs/guide/vec_envs.rst]: "Refactored" made sense when the we were still converting the models to the Sklearn-like interface, I agree it is not needed anymore.

"Multi Processing" has been a tad annoying in my eyes, as I am not sure what it means, is it SubVecEnv compatible? MPI compatible? I still think we should keep it, as a lot of methods currently dont support multi environments acceleration, over multi CPU.

**Comment by AdamGleave** [comment]
I've made the suggested improvements to the documentation, thanks for the feedback.
